### ISSUE TITLE
ConnectionError: Illegal value for NAMESPACE.NAME attribute for "/interop"

### ISSUE THREAD

**Issue by andy-maier** [pywbem/pywbem#255]
One of the standard names defined in DSP1033 for the Interop namespace is `"/interop"`, i.e. with a leading slash.

OpenPegasus causes this to be raised in PyWBEM after returning a CIMError:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pywbem/server/__init__.py", line 136, in determine_interop_ns
    self._conn.EnumerateInstanceNames(test_classname, namespace=ns)
  File "pywbem/cim_operations.py", line 1117, in EnumerateInstanceNames
    **extra)
  File "pywbem/cim_operations.py", line 670, in _imethodcall
    timeout=self.timeout)
  File "pywbem/cim_http.py", line 673, in wbem_request
    raise ConnectionError(exc_str)
pywbem.exceptions.ConnectionError: CIMError: request-not-valid, PGErrorDetail: Semantic error: on line 2: Illegal value for NAMESPACE.NAME attribute
```

The reason for that seems to be that the PyWBEM client splits this namespace name as usual at the slash character, producing two NAMESPACE elements in the CIM-XML payload (this is from `conn.last_request`):

```
<CIM CIMVERSION="2.0" DTDVERSION="2.0">
  <MESSAGE ID="1001" PROTOCOLVERSION="1.0">
    <SIMPLEREQ>
      <IMETHODCALL NAME="EnumerateInstanceNames">
        <LOCALNAMESPACEPATH>
          <NAMESPACE NAME=""/>
          <NAMESPACE NAME="interop"/>
        </LOCALNAMESPACEPATH>
        <IPARAMVALUE NAME="ClassName">
          <CLASSNAME NAME="CIM_Namespace"/>
        </IPARAMVALUE>
      </IMETHODCALL>
    </SIMPLEREQ>
  </MESSAGE>
</CIM>
```

I think my first question would be: Given that `"/interop"` is one of the official Interop namespace names, how did we think it would be represented in CIM-XML?

In other words, do we have a problem in the standard, or a bug in OP?
This is with OP 2.12.0 on RHEL 6.7.

**Comment by andy-maier** [comment]
Karl agreed to take a look into whether this should be considered a bug in OP.

**Comment by andy-maier** [comment]
Discussion: Optional or mandatory for 0.9.0?

**Comment by KSchopmeyer** [comment]
DSP 0133 states
At least one Interop namespace of a WBEM server shall have one of the following standard names:
• interop (preferred)
• /interop (DEPRECATED)
• root/interop (DEPRECATED)
• /root/interop (DEPRECATED)

DEPRECATED
The use of /interop and /root/interop for the Interop namespace name, and more generally the
use of leading slash (/) characters in any namespace name have been deprecated in version 1.1 of this
profile. Older WBEM implementations may have considered the slash separator character in a CIM object
path URI to be part of the namespace name and thus exposed the namespace name (e.g., in the Name
property of CIM_Namespace) with a leading slash character. DSP0004 does not permit namespace
names to begin with a slash.
Producers of Interop namespace names should not create a leading slash (/) character in the Interop
namespace name. Consumers of Interop namespace names shall ignore a leading slash character in
Interop namespace names when processing them (e.g., for comparison or identification purposes).
DEPRECATED

Looks like everybody is wrong here.

1. Pegasus should ignore the leading / but it does not.  We will file a bug and fix ASAP.  I file d a bug (number 10103 on pegasus.

2. At the same time, as a Producer the client should remove the leading / also. So the fix for PyWBEM is to ignore the leading / if it is supplied.

**Comment by andy-maier** [comment]
the work for pywbem is then to remove a possibly present leading slash from namespace names it gets from its caller, before it puts it into the CIM-XML request.

We agreed to do this in 0.9.0.

**Comment by KSchopmeyer** [comment]
NOTE: I think that we also have to do this test in the parser since we could receive this back from a host. That is unlikly but possible.

**Comment by KSchopmeyer** [comment]
We concluded that this is more extensive than we thought.  Since we know of nobody actually having the problem, we will move this to 0.10.0 as mandatory.

**Comment by andy-maier** [comment]
In the direction from client to server, the user can specify namespaces in the following places:

* Default namespace of `WBEMConnection`
* namespace passed as argument to most `WBEMConnection` operations
* namespace in `CIMClassName` and `CIMInstanceName` objects passed as arguments to `WBEMConnection` operations

The leading slash should be removed from a namespace as part of preparing the operation request.

**Discussion point:** Should the leading slash also be removed when initializing or updating the namespace of `CIMClassName` and `CIMInstanceName` objects?

**Comment by andy-maier** [comment]
DMTF standards about namespace names:

* DSP0004 2.8.0 defines the following ABNF rule:
  `namespaceName = IDENTIFIER *( "/" IDENTIFIER )`
  but does not use that rule in the ABNF for MOF, nor does it define that this is a namespace name (the intention can be guessed from the rule name, though), nor does it define the namespace name format anywhere else. Its definition of `objectPath` refers to section 8.5 which does not define the structure of namespace name as part of that. See also issues #978 and #979.

* DSP0201 defines that a single namespace name component is defined by the NAME attribute (CIMName entity) of the NAMESPACE element, and the description of the CIMName entity refers back to "CIM element names in DSP0004" for a definition of the format of such names. It does not define a string format for namespace names, because the components of a namespace name (`IDENTIFIER` in the `namespaceName` ABNF rule in DSP0004) are represented by a sequence of NAMESPACE elements.

* DSP0200 defines in section 6.3.2 "Encoding of CIM Object Paths within HTTP Headers and Trailers" how the namespace name components are assembled into a string, and defines that the slash is used only between the components.

* DSP0207 states in section 5.4 "Namespace Name": "The <namespace> may contain the “ / ” character but shall not begin or end with a “ / ” ."

This means, these specifications agree that the string format for a namespace name does not begin or end with a slash.

**Comment by andy-maier** [comment]
Fixed in PR #1000.
